pkg/issuer/acme/setup_test.go

```diff
@@ -0,0 +1,520 @@
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountRegistrationFailed),
+					gen.SetIssuerConditionMessage(messageAccountRegistrationFailed+someErr.Error())),
+			},
+			wantsErr: true,
+		},
+		"Attempt to register ACME account returns an ACME error in range [400,500)": {
+			issuer:                     gen.IssuerFrom(baseIssuer),
+			kfsShouldBeCalled:          true,
+			kfsKey:                     rsaPrivKey,
+			removeClientShouldBeCalled: true,
+			expectedRegisteredAcc:      &acmeapi.Account{},
+			registerErr:                acmeErr450,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountRegistrationFailed),
+					gen.SetIssuerConditionMessage(messageAccountRegistrationFailed+acmeErr450.Error())),
+			},
+		},
+		"Attempt to register ACME account returns an ACME error outside of range [400,500)": {
+			issuer:                     gen.IssuerFrom(baseIssuer),
+			kfsShouldBeCalled:          true,
+			kfsKey:                     rsaPrivKey,
+			removeClientShouldBeCalled: true,
+			expectedRegisteredAcc:      &acmeapi.Account{},
+			registerErr:                acmeErr500,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountRegistrationFailed),
+					gen.SetIssuerConditionMessage(messageAccountRegistrationFailed+acmeErr500.Error())),
+			},
+			wantsErr: true,
+		},
+		"ACME account already exists, attempting to retrieve it fails with unknown error": {
+			issuer:                     gen.IssuerFrom(baseIssuer),
+			kfsShouldBeCalled:          true,
+			kfsKey:                     rsaPrivKey,
+			removeClientShouldBeCalled: true,
+			expectedRegisteredAcc:      &acmeapi.Account{},
+			registerErr:                acmeapi.ErrAccountAlreadyExists,
+			getRegErr:                  someErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountRegistrationFailed),
+					gen.SetIssuerConditionMessage(messageAccountRegistrationFailed+someErr.Error())),
+			},
+			wantsErr: true,
+		},
+		"EAB for issuer specified, but the corresponding secret is not found": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEEAB(someString, someString)),
+			kfsShouldBeCalled:          true,
+			kfsKey:                     rsaPrivKey,
+			removeClientShouldBeCalled: true,
+			eabSecretGetErr:            notFoundErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountRegistrationFailed),
+					gen.SetIssuerConditionMessage(messageAccountRegistrationFailed+notFoundErr.Error())),
+			},
+			expectedEvents: []string{
+				fmt.Sprintf("%s %s %s", corev1.EventTypeWarning, errorAccountRegistrationFailed, messageAccountRegistrationFailed+notFoundErr.Error()),
+			},
+		},
```

Comment:
:+1: But this test seems to be in the wrong order....the EAB logic comes before the account registration in the Setup code.